### ISSUE TITLE
On Windows, running app not quit when uninstalling the app but uninstallation just completed 

### ISSUE THREAD

**Issue by kafka-yu** [electron-userland/electron-builder#7493]
Hi team,

we find a strange issue: When uninstalling, the running app is not able to quit automatically. The uninstallation is completed fast. And the files are deleted except the files that are app running used.

The package below config to pack the app, and with a custom NSIS script to do "customInstall" and "customUninstall" steps.

```
  nsis: {
         include: "./scripts/nsis.custom.installer.nsh",
        uninstallDisplayName: "Test App",
        allowToChangeInstallationDirectory: true,
        createStartMenuShortcut: true,
        runAfterFinish: true,
        oneClick: false,
        warningsAsErrors: false, // delete registry key may show warning, ignore it
        differentialPackage: false,
      },
      win: {
        icon: "win.ico",
        certificateFile: certification.certificate,
        certificatePassword: certification.password,
        signingHashAlgorithms: [certification.algorithm],
        rfc3161TimeStampServer: certification.rfcTimestamp,
        protocols: appConfig.protocols,
      },
```

The installer.nsh file content:

```
!define SCHEMENAME "test-app-open-plugin" #URL Protocol

!macro customInstall
  DetailPrint "Register ${SCHEMENAME} URI Handler"
  DeleteRegKey HKCU "Software\Classes\${SCHEMENAME}"
  WriteRegStr HKCU "Software\Classes\${SCHEMENAME}" "" "${PRODUCT_NAME}"
  WriteRegStr HKCU "Software\Classes\${SCHEMENAME}" "URL Protocol" ""
  WriteRegStr HKCU "Software\Classes\${SCHEMENAME}\DefaultIcon" "" "$INSTDIR\${APP_EXECUTABLE_FILENAME}"
  WriteRegStr HKCU "Software\Classes\${SCHEMENAME}\shell" "" ""
  WriteRegStr HKCU "Software\Classes\${SCHEMENAME}\shell\Open" "" ""
  WriteRegStr HKCU "Software\Classes\${SCHEMENAME}\shell\Open\command" "" "$INSTDIR\${APP_EXECUTABLE_FILENAME} %1"
!macroend

!macro customUnInstall
  DeleteRegKey HKCU "Software\Classes\${SCHEMENAME}"
!macroend

```

Does anyone know the reason?

* **Electron-Builder Version**:  23.6.0

* **Node Version**: 

* **Electron Version**: 20.0.0
* **Electron Type (current, beta, nightly)**: current

* **Target**: nsis, Windows

**Comment by mmaietta** [comment]
What electron-updater version are you using?

**Comment by kafka-yu** [comment]
> What electron-updater version are you using? Was this working previously?

nope, checked, the version of electron-updater is at: 5.3.0

**Comment by kafka-yu** [comment]
> What electron-updater version are you using? Was this working previously?

the version of electron-updater is at: ^5.3.0. Yeah it worked previously, Recently we add customUninstall nsh script to delete Registry Keys, but it doesn't work even this part is reverted.

**Comment by kafka-yu** [comment]
do you have any other insights?

**Comment by mmaietta** [comment]
I don't know enough about `nsh` scripting (or windows development) to be able to provide insights there.
My best advice is that this `warningsAsErrors: false` is hiding an issue with the new uninstaller script

**Comment by kafka-yu** [comment]
hmm,, I tried removing that customized nsis stuff, and in a clean Windows VM, but still no luck.

**Comment by kafka-yu** [comment]
Is there any way to enable logs on windows? I checked some issues e.g. https://github.com/brimdata/zui/issues/2713 but seems not solution.. Maybe I should upgrade to 24.1.2 to have a try?

**Comment by mmaietta** [comment]
You'll need to provide your own custom nsis binary that supports debug logging. (I tried to build it in directly to electron-builder, but 3.0.4.2 nsis was throwing off alerts from antivirus programs, so we are stuck with 3.0.4.1

https://github.com/electron-userland/electron-builder/blob/master/packages/app-builder-lib/src/targets/nsis/nsisOptions.ts#L55
```
/**
   * Allows you to provide your own `makensis`, such as one with support for debug logging via LogSet and LogText. (Logging also requires option `debugLogging = true`)
   */
  readonly customNsisBinary?: CustomNsisBinary | null
```

Quick link to the interface below:
https://github.com/electron-userland/electron-builder/blob/ac9c33ce0ffec43ec6cb71053b2177a5b872cb48/packages/app-builder-lib/src/targets/nsis/nsisOptions.ts#L4

**Comment by kafka-yu** [comment]
Thanks @mmaietta I got> "Error: LogSet specified, NSIS_CONFIG_LOG not defined." when I set

``` js
 {
 url: 'https://github.com/electron-userland/electron-builder-binaries/releases/download',
  debugLogging: true,
  checksum:
    'VKMiizYdmNdJOWpRGz4trl4lD++BvYP2irAXpMilheUP0pc93iKlWAoP843Vlraj8YG19CVn0j+dCo/hURz9+Q==',
  version: '3.0.4.1',
}
```

**Comment by mmaietta** [comment]
https://github.com/electron-userland/electron-builder/pull/6024/files
Uses `3.0.4.2` makensis which has debug compiler flag available, but the change had to be reverted due to the anti-virus issue mentioned earlier.
`3.0.4.1` does not have the debug compiler flag

**Comment by kafka-yu** [comment]
Here is the log file, would you pls help a look? Thanks.

The error I saw is:
> Delete: DeleteFile failed("C:\Users\test\AppData\Local\Programs\Test app plugin\Test app plugin.exe")

Looks like the app not quit before deleting..


[install.log](https://github.com/electron-userland/electron-builder/files/11182383/install.log)

**Comment by kafka-yu** [comment]
@mmaietta would you pls help check the above logs? thanks.

**Comment by kafka-yu** [comment]
@mmaietta  I found something that, If I changed the product name to another one, then it works.

E.g. Now it is named "Desktop Plugin", but if changed to name ("RingCentral Desktop Plugin"), it is not working. Is there a bug in the process name comparing? Since if I changed to electron-builder: 22.11.11, both work.

![image](https://user-images.githubusercontent.com/3942409/233843073-467e7622-bc90-4e32-9216-af1db2b13b4a.png)

**Comment by kafka-yu** [comment]
@indutny do you have any idea about this issue? Would you please help check? Since I saw last last time you did some changes on uninstaller part. Appricated. Thanks!

**Comment by unzld** [comment]
@kafka-yu could you provide the `customNsisBinary` setting that you used to enable logging?

**Comment by lwintermelon** [comment]
> @indutny @lwintermelon do you have any idea about this issue? Would you please help check? Since I saw last last time you did some changes on uninstaller part. Appricated. Thanks!

Sorry for the late reply. The issus is very similar to what I fixed on #6100, maybe you can check this PR and try to fix the issue.

**Comment by kafka-yu** [comment]
thanks for your reply!  @lwintermelon   I checked but have no idea about the fixing solution, could you pls give some hints about this part? Thank you!

**Comment by repl-minatee-p** [comment]
I am also facing the same issue after I changed the product name. Is there any update on this issue ?

Electron-Builder Version : 23.6.0
Electron-Updater Version: 5.3.0
Electron Version: 24.6.5

**Comment by Quincysacc** [comment]
I am having this issue. I uninstalled it, and i kept getting a popup saying, "electron.exe System error" something about dll, and how I had to reinstall the app to fix the issue. But seeing as I'd get this popup every time I turned on my PC. I knew it was running in the background causing the popups. For me what fixed it, I just opened task manager and found that electron was somehow still running in the background, I just selected it after closing the popup, and deleted it from the task manager menu. Hopefully this helps someone else.
